render/render.js

```diff
@@ -161,86 +161,215 @@ module.exports = function($window) {
+	//
+	// We will:
+	//
+	// 1. describe its general structure
+	// 2. focus on the diff algorithm optimizations
+	// 3. describe how the recycling pool meshes into this
+	// 4. discuss DOM node operations.
+
+	// ## Overview:
+	//
+	// The updateNodes() function:
+	// - deals with trivial cases
+	// - determines whether the lists are keyed or unkeyed
+	//   (we take advantage of the fact that mixed diff is not supported and settle on the
+	//   keyedness of the first vnode we find)
+	// - diffs them and patches the DOM if needed (that's the brunt of the code)
+	// - manages the leftovers: after diffing, are there:
+	//   - old nodes left to remove?
+	// 	 - new nodes to insert?
+	//   - nodes left in the recycling pool?
+	// 	 deal with them!
+	//
+	// The lists are only iterated over once, with an exception for the nodes in `old` that
+	// are visited in the fourth part of the diff and in the `removeNodes` loop.
+
+	// ## Diffing
+	//
+	// There's first a simple diff for unkeyed lists of equal length that eschews the pool.
+	//
+	// It is followed by a small section that activates the recycling pool if present, we'll
+	// ignore it for now.
+	//
+	// Then comes the main diff algorithm that is split in four parts (simplifying a bit).
+	//
+	// The first part goes through both lists top-down as long as the nodes at each level have
+	// the same key. This is always true for unkeyed lists that are entirely processed by this
+	// step.
+	//
+	// The second part deals with lists reversals, and traverses one list top-down and the other
+	// bottom-up (as long as the keys match).
+	//
+	// The third part goes through both lists bottom up as long as the keys match.
+	//
+	// The first and third sections allow us to deal efficiently with situations where one or
+	// more contiguous nodes were either inserted into, removed from or re-ordered in an otherwise
+	// sorted list. They may reduce the number of nodes to be processed in the fourth section.
+	//
+	// The fourth section does keyed diff for the situations not covered by the other three. It
+	// builds a {key: oldIndex} dictionary and uses it to find old nodes that match the keys of
+	// new ones.
+	// The nodes from the `old` array that have a match in the new `vnodes` one are marked as
+	// `vnode.skip: true`.
+	//
+	// If there are still nodes in the new `vnodes` array that haven't been matched to old ones,
+	// they are created.
+	// The range of old nodes that wasn't covered by the first three sections is passed to
+	// `removeNodes()`. Those nodes are removed unless marked as `.skip: true`.
+	//
+	// Then some pool business happens.
+	//
+	// It should be noted that the description of the four sections above is not perfect, because those
+	// parts are actually implemented as only two loops, one for the first two parts, and one for
+	// the other two. I'm not sure it wins us anything except maybe a few bytes of file size.
```

Comment:
Note for later: we should investigate whether splitting up the loops would actually help us performance-wise. Sometimes, in simpler cases, we could get some serious wins out of multiple smaller loops, but not always.

(Nothing actually needs to change until we do some benchmarking.)

Reply:
I've been researching diff algos, and studying ivi in particular since it is performant and has a nice explanation of what it does to minimize the amount of DOM operations. I've also chatted a bit with @localvoid about this, he's been really helpful.

We'll have to split the keyed/unkeyed diff completely.

In keyed mode, you must tag each node in the new list with the position of its previous incarnation in the `old` list, marking new nodes as `-1`. So going from `a b c d` to `d b a c` gives us `4 2 1 3`. In that list, the longest increasing sequence (LIS, `1 3`, so `a c`) must be left untouched, and the other nodes must be moved. There's a general algo to determine the LIS that is `O(N log(N))`, but there strategies that allow to defer generating the sequence and to narrow down the length of the sublists on which it must be computed:

As long as the keys match, a linear search from both ends can update the nodes in place, they are guaranteed to be part of the LIS and must not be moved. Also, if the nodes under the cursors from both ends can be swapped (which happens in list reversals, or if only two nodes have been swapped in an otherwise untouched list for example) they are guaranteed not to be in the LIS and they can be swapped safely. In those scenarios, the pool can be ignored since there are no nodes insertions or deletions either. 

When that strategy fails you have to generate the key=>position map and the LIS, based on the parts of the lists that still have to be processed, to determine what node has been added/removed, and which one must be moved in the DOM.

There's another tricky issue: fragments. ivi doesn't support them, but we do. If we compute the LIS based on the vnodes, we can end up moving extra DOM nodes. 

`a b ccc d` => `ccc a b d`: `333 1 2 4`, where `ccc` is a three element fragment. The vnode LIS is `1 2 4`, but if you look at DOM nodes if it is `333 4`.

 We can probably adapt the LIS algo to take the domLength into account but it will be original research AFAIK. Also users may expect to see the fragment move rather than the  other two nodes so maybe we could simply ignore the issue (that's what I plan to do initially, actually).

Reply:
To clarify, I meant splitting it into *more* than two. Sorry for the ambiguity.

Reply:
No problem, I will have to be revamped one way or another. I just wanted to dump those things I've had on my mind, and it seemed vaguely on topic here ;-)